diff -uNr ffmpeg-4.4.4/libavcodec/allcodecs.c ffmpeg-4.4.4.b/libavcodec/allcodecs.c
--- ffmpeg-4.4.4/libavcodec/allcodecs.c	2025-02-07 10:59:18.098232829 +0800
+++ ffmpeg-4.4.4.b/libavcodec/allcodecs.c	2025-02-07 10:59:41.862159108 +0800
@@ -925,7 +925,8 @@
     const AVCodec *p, *experimental = NULL;
     void *i = 0;
 
-    //Use hw codec for H264 and JPEG
+    id = remap_deprecated_codec_id(id);
+
     if(id == AV_CODEC_ID_H264){
         experimental = find_codec_by_name("h264_v4l2m2m", x);
     }
@@ -937,8 +938,6 @@
         return (AVCodec*)experimental;
     }
 
-    id = remap_deprecated_codec_id(id);
-
     while ((p = av_codec_iterate(&i))) {
         if (!x(p))
             continue;
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_buffers.c ffmpeg-4.4.4.b/libavcodec/v4l2_buffers.c
--- ffmpeg-4.4.4/libavcodec/v4l2_buffers.c	2023-04-02 23:21:32.000000000 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_buffers.c	2025-02-07 10:59:41.862159108 +0800
@@ -34,6 +34,13 @@
 #include "v4l2_buffers.h"
 #include "v4l2_m2m.h"
 
+/****same with v4l2.c*****/
+struct buff_data {
+    struct video_data *s;
+    int index;
+    int v4l2_fd;
+};
+
 #define USEC_PER_SEC 1000000
 static AVRational v4l2_timebase = { 1, USEC_PER_SEC };
 
@@ -297,6 +304,78 @@
     return 0;
 }
 
+static int v4l2_bufref_to_buf_dma(V4L2Buffer *out, void* input_opaque, int size, int offset)
+{
+    unsigned int length;
+    int *data_ptr;
+    V4L2Context *ctx = out->context;
+    V4L2m2mPriv *priv = ctx->priv;
+    struct buff_data* buf_descriptor = input_opaque;
+    
+    struct v4l2_exportbuffer expbuf = {
+        .type = V4L2_BUF_TYPE_VIDEO_CAPTURE,
+        .plane = 0,
+        // .flags = O_CLOEXEC,
+    };
+    
+    struct v4l2_streamparm parm = {
+        .type = V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE,
+	};
+
+    if(!priv || !priv->use_dma_copy)
+        return -1;
+   
+    if(!buf_descriptor || !buf_descriptor->v4l2_fd)
+        return -1;
+
+    if(out->buf.type != V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE)
+        return -1;
+    
+    if(ctx->exp_fd[VIDEO_MAX_FRAME] == 0){
+        struct v4l2_capability cap;
+        ioctl(buf_descriptor->v4l2_fd, VIDIOC_QUERYCAP, &cap);
+
+        ctx->exp_fd[VIDEO_MAX_FRAME] = strcmp("sp_vin", cap.driver) ? -1 : 1;
+
+        if(ctx->exp_fd[VIDEO_MAX_FRAME] == -1){
+            av_log(logger(out), AV_LOG_WARNING, "DMA copy not support for %s\n", cap.driver);
+        }
+    }
+
+    if(ctx->exp_fd[VIDEO_MAX_FRAME] == -1)
+        return -1;
+
+    /* get the export fd of the dma buffer, release them when the video codec driver close */
+    if(!ctx->exp_fd[buf_descriptor->index]){
+        expbuf.index = buf_descriptor->index;
+        if (ioctl(buf_descriptor->v4l2_fd, VIDIOC_EXPBUF, &expbuf) < 0) {
+            av_log(logger(out), AV_LOG_ERROR, "expbuf failed\n");
+            return -1;
+        }
+        ctx->exp_fd[buf_descriptor->index] = expbuf.fd;
+    }
+
+    length = out->plane_info[0].length;
+
+    parm.parm.raw_data[99] = 0xf1;   //cmd, 1 bytes
+
+    data_ptr = (int*)&parm.parm.raw_data[100];
+    data_ptr[0] = buf_descriptor->index; //src buf idx
+    data_ptr[1] = ctx->exp_fd[buf_descriptor->index]; //src buf fd
+    data_ptr[2] = out->buf.index;  //dst buf idx
+    data_ptr[3] = FFMIN(size, length-offset); //size
+
+    // printf("u %d %d\n", out->exp_fd[buf_descriptor->index], out->buf.index);
+
+    /* let kernel to do dma copy */
+    if (ioctl(priv->context->fd, VIDIOC_S_PARM, &parm) < 0){
+        av_log(logger(out), AV_LOG_ERROR, "DMA copy failed, use memcpy!\n");
+        return -1;
+    }
+
+    return 0;
+}
+
 static int v4l2_buffer_buf_to_swframe(AVFrame *frame, V4L2Buffer *avbuf)
 {
     int i, ret;
@@ -386,7 +465,12 @@
                 h = AV_CEIL_RSHIFT(h, desc->log2_chroma_h);
             }
             size = frame->linesize[i] * h;
-            ret = v4l2_bufref_to_buf(out, 0, frame->data[i], size, offset);
+
+            /* use dma copy if the input v4l2 dma, such as sensor */
+            if(ret = v4l2_bufref_to_buf_dma(out, av_buffer_pool_buffer_get_buffer_opaque(frame->buf[i]), size, offset)){
+                ret = v4l2_bufref_to_buf(out, 0, frame->data[i], size, offset);
+            }
+
             if (ret)
                 return ret;
             offset += size;
@@ -434,7 +518,7 @@
     frame->color_range = v4l2_get_color_range(avbuf);
     frame->color_trc = v4l2_get_color_trc(avbuf);
     frame->pts = v4l2_get_pts(avbuf);
-    frame->pkt_dts = AV_NOPTS_VALUE;
+    frame->pkt_pts = frame->pts;
 
     /* these values are updated also during re-init in v4l2_process_driver_event */
     frame->height = avbuf->context->height;
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_context.c ffmpeg-4.4.4.b/libavcodec/v4l2_context.c
--- ffmpeg-4.4.4/libavcodec/v4l2_context.c	2025-02-07 10:59:06.246269596 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_context.c	2025-02-07 10:59:41.862159108 +0800
@@ -713,6 +713,16 @@
 
 int ff_v4l2_context_set_format(V4L2Context* ctx)
 {
+    V4L2m2mPriv *priv = ctx->priv;
+    struct v4l2_format format;
+
+    if(priv->remap_sync){
+        format.fmt.raw_data[0] = 0xf1;
+        format.fmt.raw_data[1] = 0x01;
+        format.type = ctx->format.type;
+        ioctl(ctx_to_m2mctx(ctx)->fd, VIDIOC_S_FMT, &format);
+    }
+
     return ioctl(ctx_to_m2mctx(ctx)->fd, VIDIOC_S_FMT, &ctx->format);
 }
 
@@ -771,6 +781,8 @@
         }
     }
 
+    ctx->done = 0;
+
     av_log(logger(ctx), AV_LOG_DEBUG, "%s: %s %02d buffers initialized: %04ux%04u, sizeimage %08u, bytesperline %08u\n", ctx->name,
         V4L2_TYPE_IS_MULTIPLANAR(ctx->type) ? av_fourcc2str(ctx->format.fmt.pix_mp.pixelformat) : av_fourcc2str(ctx->format.fmt.pix.pixelformat),
         req.count,
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_context.h ffmpeg-4.4.4.b/libavcodec/v4l2_context.h
--- ffmpeg-4.4.4/libavcodec/v4l2_context.h	2023-04-13 02:01:50.000000000 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_context.h	2025-02-07 10:59:41.866159096 +0800
@@ -92,6 +92,12 @@
      */
     int done;
 
+    /* reference back to V4L2m2mPriv */
+    void *priv;
+
+    /* export fd if the input is from v4l2 buffer */ 
+    int exp_fd[VIDEO_MAX_FRAME+1];
+
 } V4L2Context;
 
 /**
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_m2m.c ffmpeg-4.4.4.b/libavcodec/v4l2_m2m.c
--- ffmpeg-4.4.4/libavcodec/v4l2_m2m.c	2025-02-07 10:59:06.246269596 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_m2m.c	2025-02-07 10:59:41.866159096 +0800
@@ -66,6 +66,7 @@
     void *log_ctx = s->avctx;
     int ret;
 
+    s->capture.priv = s->output.priv = s->priv;
     s->capture.done = s->output.done = 0;
     s->capture.name = "capture";
     s->output.name = "output";
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_m2m.h ffmpeg-4.4.4.b/libavcodec/v4l2_m2m.h
--- ffmpeg-4.4.4/libavcodec/v4l2_m2m.h	2025-02-07 10:59:11.302253911 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_m2m.h	2025-02-07 11:00:39.029117722 +0800
@@ -78,6 +78,13 @@
     int num_capture_buffers;
 
     int spspps_to_idr;
+    int use_dma_copy;
+
+    int remap_sync;
+
+    int reset_i;
+
+    int fixed_qp;
 } V4L2m2mPriv;
 
 /**
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_m2m_dec.c ffmpeg-4.4.4.b/libavcodec/v4l2_m2m_dec.c
--- ffmpeg-4.4.4/libavcodec/v4l2_m2m_dec.c	2025-02-07 10:59:06.246269596 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_m2m_dec.c	2025-02-07 10:59:41.866159096 +0800
@@ -197,7 +197,14 @@
     }
 
 dequeue:
-    return ff_v4l2_context_dequeue_frame(capture, frame, -1);
+    ret = ff_v4l2_context_dequeue_frame(capture, frame, -1);
+
+    /* if get nonezero packet when decode EOS, reinit the codec to restart. */
+    if(ret == AVERROR_EOF && s->buf_pkt.size){
+        ff_v4l2_m2m_codec_full_reinit(s);
+        return AVERROR(EAGAIN);
+    }
+    return ret;
 fail:
     av_packet_unref(&s->buf_pkt);
     return ret;
@@ -253,6 +260,8 @@
     V4L_M2M_DEFAULT_OPTS,
     { "num_capture_buffers", "Number of buffers in the capture context",
         OFFSET(num_capture_buffers), AV_OPT_TYPE_INT, {.i64 = 20}, 20, INT_MAX, FLAGS },
+    { "remap_sync", "Improve video buffer access speed by cache DMA to CPU", 
+        OFFSET(remap_sync), AV_OPT_TYPE_BOOL, {.i64 = 1 }, 0, 1, FLAGS },
     { NULL},
 };
 
diff -uNr ffmpeg-4.4.4/libavcodec/v4l2_m2m_enc.c ffmpeg-4.4.4.b/libavcodec/v4l2_m2m_enc.c
--- ffmpeg-4.4.4/libavcodec/v4l2_m2m_enc.c	2025-02-07 10:59:11.302253911 +0800
+++ ffmpeg-4.4.4.b/libavcodec/v4l2_m2m_enc.c	2025-02-07 11:01:35.323110470 +0800
@@ -188,6 +188,7 @@
 static int v4l2_prepare_encoder(V4L2m2mContext *s)
 {
     AVCodecContext *avctx = s->avctx;
+    V4L2m2mPriv *priv = avctx->priv_data;
     int qmin_cid, qmax_cid, qmin, qmax;
     int ret, val;
 
@@ -279,6 +280,16 @@
         qmin = 0;
         qmax = 255;
         break;
+    case AV_CODEC_ID_MJPEG:
+        struct v4l2_control ctrl = {0};
+
+        ctrl.id = V4L2_CID_JPEG_COMPRESSION_QUALITY;
+        ctrl.value = 101 - (priv->fixed_qp - 1) * 2;
+        
+        if(ioctl(s->fd, VIDIOC_S_CTRL, &ctrl) < 0){
+            av_log(s->avctx, AV_LOG_WARNING, "Failed to set qp level = %d %s\n", priv->fixed_qp, strerror(errno));
+        }
+        break;
     default:
         return 0;
     }
@@ -320,6 +331,7 @@
     }
 #endif
 
+    s->output.priv = avctx->priv_data;
     return ff_v4l2_context_enqueue_frame(output, frame);
 }
 
@@ -344,6 +356,12 @@
             frame = NULL;
     }
 
+    if(priv->reset_i){
+        v4l2_set_ext_ctrl(s, MPEG_CID(FORCE_KEY_FRAME), 1, "force key frame", 1);
+
+        av_log(avctx, AV_LOG_INFO, "Set FORCE_KEY_FRAME\n");
+    }
+
     ret = v4l2_send_frame(avctx, frame);
     if (ret != AVERROR(EAGAIN))
         av_frame_unref(frame);
@@ -461,7 +479,11 @@
 #define V4L_M2M_CAPTURE_OPTS \
     V4L_M2M_DEFAULT_OPTS,\
     { "num_capture_buffers", "Number of buffers in the capture context", \
-        OFFSET(num_capture_buffers), AV_OPT_TYPE_INT, {.i64 = 4 }, 4, INT_MAX, FLAGS }
+        OFFSET(num_capture_buffers), AV_OPT_TYPE_INT, {.i64 = 4 }, 4, INT_MAX, FLAGS }, \
+    { "use_dma_copy", "Use dma copy when video buffer is from DMA", \
+        OFFSET(use_dma_copy), AV_OPT_TYPE_BOOL, {.i64 = 1 }, 0, 1, FLAGS }, \
+    { "remap_sync", "Improve video buffer access speed by cache DMA to CPU", \
+        OFFSET(remap_sync), AV_OPT_TYPE_BOOL, {.i64 = 1 }, 0, 1, FLAGS }
 
 static const AVOption mpeg4_options[] = {
     V4L_M2M_CAPTURE_OPTS,
@@ -469,10 +491,21 @@
     { NULL },
 };
 
+static const AVOption h264_options[] = {
+    V4L_M2M_CAPTURE_OPTS,
+    { "spspps_to_idr", "Add spspps to each key frame", OFFSET(spspps_to_idr), AV_OPT_TYPE_BOOL, {.i64 = 1 }, 0, 1, FLAGS },
+    { "reset_i", "Force to key frame", OFFSET(reset_i), AV_OPT_TYPE_BOOL, {.i64 = 0 }, 0, 1, FLAGS },
+    { NULL },
+};
+
+static const AVOption jpeg_options[] = {
+    V4L_M2M_CAPTURE_OPTS,
+    { "qp", "Set qp level to jpeg encode", OFFSET(fixed_qp), AV_OPT_TYPE_INT, {.i64 = 3 }, 0, 10, FLAGS },
+    { NULL },
+};
+
 static const AVOption options[] = {
     V4L_M2M_CAPTURE_OPTS,
-    { "spspps_to_idr", "Add spspps to each key frame", \
-	OFFSET(spspps_to_idr), AV_OPT_TYPE_BOOL, {.i64 = 1 }, 0, 1, FLAGS },
     { NULL },
 };
 
@@ -510,7 +543,7 @@
 
 M2MENC(mpeg4,"MPEG4", mpeg4_options, AV_CODEC_ID_MPEG4);
 M2MENC(h263, "H.263", options,       AV_CODEC_ID_H263);
-M2MENC(h264, "H.264", options,       AV_CODEC_ID_H264);
+M2MENC(h264, "H.264", h264_options,       AV_CODEC_ID_H264);
 M2MENC(hevc, "HEVC",  options,       AV_CODEC_ID_HEVC);
 M2MENC(vp8,  "VP8",   options,       AV_CODEC_ID_VP8);
-M2MENC(jpeg, "JPEG",   options,     AV_CODEC_ID_MJPEG);
+M2MENC(jpeg, "JPEG",  jpeg_options,     AV_CODEC_ID_MJPEG);
diff -uNr ffmpeg-4.4.4/libavcodec/vaapi_h264.c ffmpeg-4.4.4.b/libavcodec/vaapi_h264.c
--- ffmpeg-4.4.4/libavcodec/vaapi_h264.c	2025-02-07 10:58:44.934335711 +0800
+++ ffmpeg-4.4.4.b/libavcodec/vaapi_h264.c	2025-02-07 10:59:41.866159096 +0800
@@ -314,12 +314,12 @@
     H264SliceContext *sl = &h->slice_ctx[0];
     int ret;
 
-    if (pic->nb_slices == 0) {
-        ret = AVERROR_INVALIDDATA;
+	if (pic->nb_slices == 0) {
+	    ret = AVERROR_INVALIDDATA;
         goto finish;
     }
 
-    ret = ff_vaapi_decode_issue(avctx, pic);
+	ret = ff_vaapi_decode_issue(avctx, pic);
     if (ret < 0)
         goto finish;
 
diff -uNr ffmpeg-4.4.4/libavdevice/v4l2.c ffmpeg-4.4.4.b/libavdevice/v4l2.c
--- ffmpeg-4.4.4/libavdevice/v4l2.c	2023-04-13 02:01:50.000000000 +0800
+++ ffmpeg-4.4.4.b/libavdevice/v4l2.c	2025-02-07 10:59:41.866159096 +0800
@@ -108,6 +108,7 @@
 struct buff_data {
     struct video_data *s;
     int index;
+    int v4l2_fd;
 };
 
 static int device_open(AVFormatContext *ctx, const char* device_path)
@@ -579,6 +580,7 @@
         }
         buf_descriptor->index = buf.index;
         buf_descriptor->s     = s;
+        buf_descriptor->v4l2_fd = s->fd;
 
         pkt->buf = av_buffer_create(pkt->data, pkt->size, mmap_release_buffer,
                                     buf_descriptor, 0);
diff -uNr ffmpeg-4.4.4/libavutil/buffer.c ffmpeg-4.4.4.b/libavutil/buffer.c
--- ffmpeg-4.4.4/libavutil/buffer.c	2023-04-13 02:01:50.000000000 +0800
+++ ffmpeg-4.4.4.b/libavutil/buffer.c	2025-02-07 10:59:41.866159096 +0800
@@ -401,3 +401,9 @@
     av_assert0(buf);
     return buf->opaque;
 }
+
+void *av_buffer_pool_buffer_get_buffer_opaque(AVBufferRef *ref){
+    if(!ref) return NULL;
+    AVBuffer* buf = ref->buffer;
+    return buf ? buf->opaque : NULL;
+}
diff -uNr ffmpeg-4.4.4/libavutil/buffer.h ffmpeg-4.4.4.b/libavutil/buffer.h
--- ffmpeg-4.4.4/libavutil/buffer.h	2023-04-13 02:01:50.000000000 +0800
+++ ffmpeg-4.4.4.b/libavutil/buffer.h	2025-02-07 10:59:41.866159096 +0800
@@ -347,6 +347,11 @@
 void *av_buffer_pool_buffer_get_opaque(AVBufferRef *ref);
 
 /**
+ * Get the AVBuffer opaque parameter of the AVBufferRef.
+ */
+void *av_buffer_pool_buffer_get_buffer_opaque(AVBufferRef *ref);
+
+/**
  * @}
  */
 
